File: src-tauri/src/stats/syncer.rs
use crate::db::{
    CohortAccounts, CohortFrames, Cohorts,
    Frames,
};
use crate::ssh::SSH;
use crate::stats::fetcher::StatsFetcher;
use crate::stats::structs::{IEarningsFileCohort};
use crate::config::{ConfigFile, ServerDetails};
use anyhow::Result;
use log::{error, info};
use std::collections::HashSet;
use std::str::FromStr;
use std::sync::Mutex;
use lazy_static::lazy_static;

lazy_static! {
    static ref IS_RUNNING: Mutex<Option<tokio::task::JoinHandle<()>>> = Mutex::new(None);
}

pub struct StatsSyncer {
    ssh: SSH,
    local_port: u16,
}

impl StatsSyncer {
    pub async fn new(ssh: &SSH, local_port: u16) -> Result<Self> {
        let ssh = SSH::connect(ssh.config.clone()).await?;

        Ok(Self {
            ssh: ssh.clone(),
            local_port,
        })
    }

    pub async fn launch_sync_thread(ssh: &SSH, local_port: u16, oldest_frame_id_to_sync: u32, current_frame_id: u32) -> Result<()> {       
        if let Some(_handle) = IS_RUNNING.lock().unwrap().take() {
            info!("Syncer thread already running, skipping");
            return Ok(());
        }
        let mut syncer = StatsSyncer::new(ssh, local_port).await?;

        let handle: tokio::task::JoinHandle<()> = tokio::spawn(async move {
            match syncer.sync_db(oldest_frame_id_to_sync, current_frame_id).await {
                Ok(_) => info!("Syncer thread Finished"),
                Err(e) => {
                    error!("Syncer thread failed: {}", e);
                    let mut server_details = ServerDetails::load().unwrap_or_default();
                    server_details.sync_error = Some(e.to_string());
                    server_details.save().unwrap_or_default();
                }
            }
            let _ = IS_RUNNING.lock().unwrap().take();
        });

        let _ = IS_RUNNING.lock().unwrap().insert(handle);

        Ok(())
    }

    async fn sync_db(&mut self, oldest_frame_id_to_sync: u32, current_frame_id: u32) -> Result<()> {
        for frame_id in oldest_frame_id_to_sync..=current_frame_id {
            info!("Syncing frame: {}", frame_id);
            self.sync_db_frame(frame_id).await?;
        }

        Ok(())
    }

    pub async fn sync_db_frame(&mut self, frame_id: u32) -> Result<()> {
        if let Ok(frame) = Frames::fetch_by_id(frame_id) {
            if frame.is_processed {
                return Ok(());
            }
        }

        let earnings_file = StatsFetcher::fetch_earnings_file(self.local_port, frame_id).await?;

        Frames::insert_or_update(
            frame_id,
            earnings_file.frame_tick_start,
            earnings_file.frame_tick_end,
            earnings_file.frame_progress,
            false,
        )?;

        let mut processed_cohorts: HashSet<u32> = HashSet::new();

        let mut total_blocks_mined = 0;

        for (cohort_frame_id_str, cohort_data) in
            earnings_file.by_cohort_frame_id.iter()
        {
            let cohort_frame_id = u32::from_str(cohort_frame_id_str)
                .map_err(|e| anyhow::anyhow!("Failed to parse frame ID: {}", e))?;
            if !processed_cohorts.contains(&cohort_frame_id) {
                self.sync_db_cohort(cohort_frame_id, frame_id, earnings_file.frame_progress)
                    .await?;
                processed_cohorts.insert(cohort_frame_id);
            }

            total_blocks_mined += cohort_data.blocks_mined;
            self.sync_db_cohort_frame(cohort_frame_id, frame_id, cohort_data)
                .await?;
        }

        info!("TOTAL BLOCKS MINED: {}", total_blocks_mined);

        let is_processed = earnings_file.frame_progress == 100.0;
        Frames::update(
            frame_id,
            earnings_file.frame_tick_start,
            earnings_file.frame_tick_end,
            earnings_file.frame_progress,
            is_processed,
        )?;

        Ok(())
    }

    async fn sync_db_cohort_frame(
        &mut self,
        cohort_bidding_frame_id: u32,
        frame_id: u32,
        data: &IEarningsFileCohort,
    ) -> Result<()> {
        CohortFrames::insert_or_update(
            frame_id,
            cohort_bidding_frame_id,
            data.blocks_mined,
            data.argonots_mined,
            data.argons_mined,
            data.argons_minted,
        )?;

        Ok(())
    }

    async fn sync_db_cohort(
        &mut self,
        cohort_starting_frame_id: u32,
        current_frame_id: u32,
        current_frame_progress: f32,
    ) -> Result<()> {
        let data = StatsFetcher::fetch_bids_file(self.local_port, Some(cohort_starting_frame_id)).await?;

        if data.frame_bidding_progress < 100.0 {
            return Ok(());
        }

        let frames_completed = current_frame_id - cohort_starting_frame_id;
        let progress = ((frames_completed as f32 * 100.0) + current_frame_progress) / 10.0;
        let argonots_staked = data.argonots_staked_per_seat * data.seats_won as u64;

        Cohorts::insert_or_update(
            cohort_starting_frame_id,
            progress,
            data.transaction_fees,
            argonots_staked,
            data.argons_bid_total,
            data.seats_won,
        )?;

        CohortAccounts::delete_for_cohort(cohort_starting_frame_id)?;

        for subaccount in data.subaccounts {
            CohortAccounts::insert(
                subaccount.index,
                cohort_starting_frame_id,
                subaccount.address,
                subaccount.argons_bid.unwrap_or_default(),
                subaccount.bid_position.unwrap_or_default(),
            )?;
        }

        Ok(())
    }
}

File: src-vue/lib/stats/Syncer.ts
import { Config } from '../Config';
import { Db } from '../Db';
import { StatsFetcher } from './Fetcher';
import { IEarningsFileCohort, ISyncState } from '@argonprotocol/commander-bot/src/storage';

export class StatsSyncer {
  public isMissingCurrentFrame: boolean = false;

  private isSynced: boolean = false;
  private db: Db;
  private config: Config;

  private oldestFrameIdToSync!: number;
  private currentFrameId!: number;

  constructor(config: Config, db: Db) {
    this.config = config;
    this.db = db;
  }

  public async sync(syncState: ISyncState): Promise<void> {
    if (this.isSynced) return;
    console.log('== SYNC STARTED ==========================================');

    this.oldestFrameIdToSync = syncState.oldestFrameIdToSync;
    this.currentFrameId = syncState.currentFrameId;

    const rawProgress = await this.calculateSyncProgress(syncState);

    if (rawProgress >= 100.0) {
      if (this.config.syncDetails.startDate) {
        this.config.resetField('syncDetails');
        this.config.save();
      }
      this.isSynced = true;
      console.log('SYNC NOT NEEDED');
      return undefined;
    }

    if (!this.config.syncDetails.startDate) {
      this.config.syncDetails = {
        progress: 0,
        startDate: new Date().toISOString(),
        startPosition: rawProgress,
        errorType: null,
        errorMessage: null,
      };
      await this.config.save();
    }

    while (syncState.loadProgress < 100.0) {
      await new Promise(resolve => setTimeout(resolve, 1000));
      syncState = await StatsFetcher.fetchSyncState(this.localPort);
      this.updateSyncDetailsProgress(syncState);
    }

    for (let frameId = this.oldestFrameIdToSync; frameId <= this.currentFrameId; frameId++) {
      console.log('Syncing frame:', frameId);
      await this.syncDbFrame(frameId);
      this.updateSyncDetailsProgress(syncState);
    }

    this.isMissingCurrentFrame = false;

    console.log('== SYNC COMPLETED ==========================================');
    this.isSynced = true;
  }

  public async syncDbFrame(frameId: number): Promise<void> {
    try {
      const frame = await this.db.framesTable.fetchById(frameId);
      if (frame.isProcessed) {
        return;
      }
    } catch (error) {
      console.log('Frame not found:', frameId);
    }

    const earningsFile = await StatsFetcher.fetchEarningsFile(frameId);

    await this.db.framesTable.insertOrUpdate(
      frameId,
      earningsFile.firstTick,
      earningsFile.lastTick,
      earningsFile.firstBlockNumber,
      earningsFile.lastBlockNumber,
      earningsFile.frameProgress,
      false,
    );

    const processedCohorts: Set<number> = new Set();
    let totalBlocksMined = 0;

    const cohortsById = Object.entries(earningsFile.byCohortActivatingFrameId) as [
      string,
      IEarningsFileCohort,
    ][];

    for (const [cohortActivatingFrameIdStr, cohortData] of cohortsById) {
      const cohortActivatingFrameId = parseInt(cohortActivatingFrameIdStr, 10);
      if (!processedCohorts.has(cohortActivatingFrameId)) {
        await this.syncDbCohort(cohortActivatingFrameId, frameId, earningsFile.frameProgress);
        processedCohorts.add(cohortActivatingFrameId);
      }

      totalBlocksMined += cohortData.blocksMined;
      await this.syncDbCohortFrame(cohortActivatingFrameId, frameId, cohortData);
    }

    const isProcessed = earningsFile.frameProgress === 100.0;
    await this.db.framesTable.update(
      frameId,
      earningsFile.firstTick,
      earningsFile.lastTick,
      earningsFile.firstBlockNumber,
      earningsFile.lastBlockNumber,
      earningsFile.frameProgress,
      isProcessed,
    );
  }

  private async syncDbCohortFrame(
    cohortActivatingFrameId: number,
    frameId: number,
    cohortData: IEarningsFileCohort,
  ): Promise<void> {
    await this.db.cohortFramesTable.insertOrUpdate(
      frameId,
      cohortActivatingFrameId,
      cohortData.blocksMined,
      Number(cohortData.argonotsMined),
      Number(cohortData.argonsMined),
      Number(cohortData.argonsMinted),
    );
  }

  private async syncDbCohort(
    cohortActivatingFrameId: number,
    currentFrameId: number,
    currentFrameProgress: number,
  ): Promise<void> {
    const data = await StatsFetcher.fetchBidsFile(cohortActivatingFrameId);

    if (data.frameBiddingProgress < 100.0) {
      return;
    }

    const framesCompleted = currentFrameId - cohortActivatingFrameId;
    const progress = (framesCompleted * 100.0 + currentFrameProgress) / 10.0;
    const argonotsStaked = Number(data.argonotsStakedPerSeat) * data.seatsWon;

    await this.db.cohortsTable.insertOrUpdate(
      cohortActivatingFrameId,
      progress,
      Number(data.transactionFees),
      argonotsStaked,
      Number(data.argonsBidTotal),
      data.seatsWon,
    );

    await this.db.cohortAccountsTable.deleteForCohort(cohortActivatingFrameId);

    for (const subaccount of data.winningBids) {
      if (subaccount.subAccountIndex === undefined) return;
      await this.db.cohortAccountsTable.insert(
        subaccount.subAccountIndex,
        cohortActivatingFrameId,
        subaccount.address,
        Number(subaccount.argonsBid ?? 0n),
        Number(subaccount.bidPosition ?? 0n),
      );
    }
  }

  private async updateSyncDetailsProgress(syncState: ISyncState): Promise<void> {
    const rawProgress = await this.calculateSyncProgress(syncState);
    const startPosition = this.config.syncDetails.startPosition ?? 0;
    const adjustedProgress = ((rawProgress - startPosition) / (100 - startPosition)) * 100;
    this.config.syncDetails.progress = adjustedProgress;
    this.config.syncDetails = this.config.syncDetails;
    await this.config.save();
  }

  private async calculateSyncProgress(syncState: ISyncState): Promise<number> {
    const botLoadProgress = syncState.loadProgress * 0.9;

    if (botLoadProgress < 90.0) {
      return botLoadProgress;
    }

    const dbSyncProgress = (await this.calculateDbSyncProgress()) * 0.1;

    return botLoadProgress + dbSyncProgress;
  }

  private async calculateDbSyncProgress(): Promise<number> {
    const dbFramesExpected = this.currentFrameId - this.oldestFrameIdToSync + 1; // +1 because we want to include the current frame
    const dbFramesProcessed = await this.db.framesTable.fetchProcessedCount();

    if (dbFramesProcessed === dbFramesExpected) {
      this.isMissingCurrentFrame = false;
      return 100.0;
    }

    if (dbFramesProcessed === dbFramesExpected - 1) {
      try {
        const currentFrame = await this.db.framesTable.fetchById(this.currentFrameId);
        if (!currentFrame.isProcessed) {
          return 100.0;
        }
      } catch {
        this.isMissingCurrentFrame = true;
        return 100.0;
      }
    }

    return (dbFramesProcessed / dbFramesExpected) * 100.0;
  }
}
